File: app/src/main/java/lazy/of/go/to/module/splash/SplashModule.java
package lazy.of.go.to.module.splash;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import dagger.android.ContributesAndroidInjector;
import lazy.of.go.to.app.splash.SplashActivity;
import lazy.of.go.to.app.splash.SplashContract;
import lazy.of.go.to.app.splash.SplashFragment;
import lazy.of.go.to.app.splash.SplashPresenter;
import lazy.of.go.to.auth.LazyAuth;
import lazy.of.go.to.auth.firebase.FbAuth;
import lazy.of.go.to.common.Log;
import lazy.of.go.to.di.ActivityScoped;
import lazy.of.go.to.di.FragmentScoped;

/**
 * @author piece.of.lazy
 */
@Module
public abstract class SplashModule {

    @FragmentScoped
    @ContributesAndroidInjector
    abstract SplashFragment injectFragment();

    @ActivityScoped
    @Binds
    abstract SplashContract.Presenter injectPresenter(SplashPresenter presenter);

    @ActivityScoped
    @Provides
    static Log provideLog(SplashActivity activity) {
        return activity.getLog();
    }

    @ActivityScoped
    @Provides
    static LazyAuth provideAuth() {
        return new FbAuth();
    }
}

File: app/src/main/kotlin/lazy/of/go/to/app/splash/SplashModule.kt
package lazy.of.go.to.app.splash

import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.android.ContributesAndroidInjector
import lazy.of.go.to.app.main.MainActivity
import lazy.of.go.to.base.feature.GetFeature
import lazy.of.go.to.base.feature.LoadingFeature
import lazy.of.go.to.common.Log
import lazy.of.go.to.db.DbInjection
import lazy.of.go.to.di.ActivityScoped
import lazy.of.go.to.di.FragmentScoped
import lazy.of.go.to.domain.data.DbUser

/**
 * @author piece.of.lazy
 */
@Module
abstract class SplashModule {
    @FragmentScoped
    @ContributesAndroidInjector
    internal abstract fun injectFragment(): SplashFragment

    @ActivityScoped
    @Binds
    internal abstract fun injectPresenter(presenter: SplashPresenter): SplashContract.Presenter

//    @ActivityScoped
//    @Binds
//    internal abstract fun injectFeatureListener(activity: MainActivity): FeatureListener

    @Module
    companion object {
        @JvmStatic
        @ActivityScoped
        @Provides
        internal fun provideLog(activity: SplashActivity): Log = activity.log

        @JvmStatic
        @ActivityScoped
        @Provides
        internal fun provideDbUser(dbInjection: DbInjection) = dbInjection.getDB(DbUser::class)

        @JvmStatic
        @ActivityScoped
        @Provides
        internal fun provideGetFeature(activity: SplashActivity): GetFeature = activity
    }
}